# ğŸ§¬ TimeVault
mRNAcap Processing Pipeline


This repository contains scripts for processing 4SU-labeled mRNA capture data, including trimming, alignment, counting, SNP calling, and identifying new reads.

## ğŸ“ Project Structure
```
MAIN_DIR/ (e.g. /data/qiyu/mRNAcap)
â”œâ”€â”€ data/
â”‚   â””â”€â”€ 250306_4SUdata_WalkUp/
â”‚       â”œâ”€â”€ submit_jobs.sh          # Job submission script from `src` 
â”‚       â”œâ”€â”€ 1_trimmed/              # Trimmed FASTQ files
â”‚       â”œâ”€â”€ 2_alignment/            # STAR alignment results
â”‚       â”œâ”€â”€ 3_bam/                  # Sorted and indexed BAM files
â”‚       â”œâ”€â”€ 4_counts/               # Count matrices
â”‚       â”œâ”€â”€ 5_SNP/                  # SNP calling results
â”‚       â”œâ”€â”€ fastq/                  # Input FASTQ files
â”‚       â””â”€â”€ logs/                   # Job logs
â”‚
â”œâ”€â”€ reference/
â”‚   â”œâ”€â”€ Genome/
â”‚   â”‚   â””â”€â”€ GRCh38/                 # Reference genome (FASTA, GTF)
â”‚   â””â”€â”€ human_STAR_index/           # STAR index files
â”‚
â””â”€â”€ src/
â”‚   â”œâ”€â”€ submit_jobs.sh           
â”‚   â”œâ”€â”€ result_check.sh          
â”‚   â”œâ”€â”€ annotation/                 # Gene annotation
â”‚   â”œâ”€â”€ config/                     # Configuration file
â”‚   â”œâ”€â”€ downsample/                 # Downsampling bam or fastq
â”‚   â””â”€â”€ process/                    # Modular processing scripts
```


## ğŸš€ Steps 
1. Clone the GitHub repository to your local directory.
2. Configure the package and reference directory as outlined in `config.sh`.
3. Copy `submit_jobs.sh` into run-specific folders, fill in your sample information, and execute it using: `bash ./submit_jobs.sh`.
4. Locate the final combined count matrix at: `4_counts/combined_all_new_counts.txt`.
